python/lsst/daf/butler/_butler.py

```diff
@@ -648,6 +648,93 @@ def _findDatasetRef(self, datasetRefOrType: Union[DatasetRef, DatasetType, str],
                                     "Could not find matching alternative (primary key has type %s).",
                                     value, dimensionName, dimension.primaryKey.getPythonType())
 
+        # If we have some unrecognized dimensions we have to try to connect
+        # them to records in other dimensions.  This is made more complicated
+        # by some dimensions having records with clashing names.  A mitigation
+        # is that we can tell by this point which dimensions are missing
+        # for the DatasetType but this does not work for calibrations
+        # where additional dimensions can be used to constrain the temporal
+        # axis.
+        if not_dimensions:
+            # Calculate missing dimensions
+            provided = set(newDataId) | set(kwds) | set(byRecord)
+            missingDimensions = datasetType.dimensions.names - provided
+
+            # For calibrations we may well be needing temporal dimensions
+            # so rather than always including all dimensions in the scan
+            # restrict things a little. It is still possible for there
+            # to be confusion over day_obs in visit vs exposure for example.
+            # If we are not searching calibration collections things may
+            # fail but they are going to fail anyway because of the
+            # ambiguousness of the dataId...
+            candidateDimensions = set()
+            candidateDimensions.update(missingDimensions)
+            if datasetType.isCalibration():
+                for dim in self.registry.dimensions.getStaticDimensions():
+                    if dim.temporal:
+                        candidateDimensions.add(str(dim))
+
+            # Look up table for the first association with a dimension
+            guessedAssociation: dict[Any, dict[str, Any]] = defaultdict(dict)
```

Comment:
Update: these annotations still don't work with my version of mypy - and they don't work with yours either, because the mypy issues is still open (https://github.com/python/mypy/issues/7907) and it turns out we're not type-checking this file at all!  We've only enabled mypy for a bunch of specific `daf.butler` subpackages, but not `_butler.py` (and we're probably missing a few other modules, too).

Reply:
I'm going to fix the mypy config and type annotations in at least this file on https://jira.lsstcorp.org/browse/DM-27689.

Reply:
Good spot. I'll leave it for you to fix then and just merge this.